app/src/main/java/org/thoughtcrime/securesms/database/ConversationNotificationDebouncer.kt

```diff
@@ -17,20 +17,28 @@ class ConversationNotificationDebouncer(private val context: Context) {
         @Synchronized
         fun get(context: Context): ConversationNotificationDebouncer {
             if (::shared.isInitialized) { return shared }
-            shared = ConversationNotificationDebouncer(context)
+            shared = ConversationNotificationDebouncer(context.applicationContext as ApplicationContext)
             return shared
         }
     }
 
     fun notify(threadID: Long) {
-        threadIDs.add(threadID)
+        synchronized(threadIDs) {
+            threadIDs.add(threadID)
+        }
+
         debouncer.publish { publish() }
     }
 
     private fun publish() {
-        for (threadID in threadIDs.toList()) {
+        val toNotify = synchronized(threadIDs) {
```

Comment:
I was just thinking, you can save the list copy by making it a `var`:

```
private fun publish() {
    synchronized(this) {
        threadIDs.also { threadIDs = mutableSetOf() }
    }.forEach {
context.contentResolver.notifyChange(DatabaseContentProviders.Conversation.getUriForThread(it), null)
}
```

but it needs the same `synchronized(this)` in `notify`.

Creating a new instance still needs an allocation, but `toList` needs an allocation and an `O(n)` copy. Probably negligible performance, so don't feel obliged to change it, but the syntax is nice-ish too.

Reply:
Also considered using a sorted `long[]` to avoid boxing but feel really unnecessary here, most of the time you only see a few items in the list really...